daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/TypeOrdering.scala

```diff
@@ -38,52 +29,72 @@ object TypeOrdering extends Ordering[Ast.Type] {
 
-  private def typeRank(typ: Ast.Type): Int =
+  private[this] def typeRank(typ: Ast.Type): Int =
     typ match {
       case Ast.TBuiltin(_) => 0
       case Ast.TTyCon(_) => 1
       case Ast.TNat(_) => 2
       case Ast.TStruct(_) => 3
       case Ast.TApp(_, _) => 4
       case Ast.TVar(_) | Ast.TForall(_, _) | Ast.TSynApp(_, _) =>
-        throw new IllegalArgumentException(s"cannot compare types $typ")
+        throw new IllegalArgumentException(s"$typ is not Comparable")
     }
 
-  @tailrec
-  // Any two ground types (types without variable nor quantifiers) can be compared.
-  private[this] def compareType(x: Int, stack0: FrontStack[(Ast.Type, Ast.Type)]): Int =
-    stack0 match {
-      case FrontStack() =>
-        x
-      case FrontStackCons(tuple, stack) =>
-        if (x != 0) x
-        else
-          tuple match {
-            case (Ast.TBuiltin(b1), Ast.TBuiltin(b2)) =>
-              compareType(builtinTypeIdx(b1) compareTo builtinTypeIdx(b2), stack)
-            case (Ast.TTyCon(con1), Ast.TTyCon(con2)) =>
-              compareType(con1 compare con2, stack)
-            case (Ast.TNat(n1), Ast.TNat(n2)) =>
-              compareType(n1 compareTo n2, stack)
-            case (Ast.TStruct(fields1), Ast.TStruct(fields2)) =>
-              compareType(
-                math.Ordering
-                  .Iterable[String]
-                  .compare(fields1.toSeq.map(_._1), fields2.toSeq.map(_._1)),
-                zipAndPush(fields1.iterator.map(_._2), fields2.iterator.map(_._2), stack)
-              )
-            case (Ast.TApp(t11, t12), Ast.TApp(t21, t22)) =>
-              compareType(0, (t11, t21) +: (t12, t22) +: stack)
-            case (t1, t2) =>
-              // This case only occurs when t1 and t2 have different ranks.
-              val x = typeRank(t1) compareTo typeRank(t2)
-              assert(x != 0)
-              x
-          }
+  // Stack safe comparison for comparable types, i.e. types that do not contain
+  // variables, quantifiers, nor synonyms.
+  @throws[IllegalArgumentException]
+  def compare(x: Ast.Type, y: Ast.Type): Int = {
+
+    var diff = 0
+    var stackX = List(Iterator.single(x))
+    var stackY = List(Iterator.single(y))
+    // invariant: stackX.length == stackY.length
+
+    @inline
+    def push(xs: Iterator[Ast.Type], ys: Iterator[Ast.Type]): Unit = {
+      stackX = xs :: stackX
+      stackY = ys :: stackY
+    }
+
+    @inline
+    def pop(): Unit = {
+      stackX = stackX.tail
+      stackY = stackY.tail
+    }
+
+    @inline
+    def step(tuple: (Ast.Type, Ast.Type)): Unit =
+      tuple match {
+        case (Ast.TBuiltin(x), Ast.TBuiltin(y)) =>
+          diff = builtinTypeRank(x) compare builtinTypeRank(y)
+        case (Ast.TTyCon(con1), Ast.TTyCon(con2)) =>
+          diff = con1 compare con2
+        case (Ast.TNat(x), Ast.TNat(y)) =>
+          diff = x compare y
+        case (Ast.TStruct(ImmArray((xField, xType))), Ast.TStruct(ImmArray((yField, yType)))) =>
```

Comment:
Maybe that’s just me but splitting multi-field structs into single-field structs seems quite convoluted and at least deserves a comment. This is also the only case where you actually make use of the fact that the elements of your stacks are iterators. Not quite sure how to clean it up, is there a reason why we wrote the spec to first compare individual field types before comparing the number of fields? The latter seems more intuitive and would also make the implementation simpler. I guess it’s still in 1.dev so if we want to change it now would be the time.

Reply:
changed.